File: advance/src/eu/advance/logistics/flow/engine/model/fd/AdvanceBlockReference.java
/*
 * Copyright 2010-2013 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */

package eu.advance.logistics.flow.engine.model.fd;

import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.common.collect.Lists;
import com.google.common.collect.Maps;

import edu.umd.cs.findbugs.annotations.NonNull;
import edu.umd.cs.findbugs.annotations.Nullable;
import eu.advance.logistics.flow.engine.util.Strings;
import eu.advance.logistics.flow.engine.xml.XElement;
import eu.advance.logistics.flow.engine.xml.XSerializable;

/**
 * The block reference for the {@code flow-description.xsd} used within composite blocks.
 * @author akarnokd, 2011.06.21.
 */
public class AdvanceBlockReference implements XSerializable {
	/** The logger. */
	protected static final Logger LOG = LoggerFactory.getLogger(AdvanceBlockReference.class);
	/** The unique identifier of this block among the current level of blocks. */
	@NonNull
	public String id;
	/** The block type identifier referencing a block description. */
	@NonNull
	public String type;
	/** The user-entered documentation of this composite block. */
	@Nullable
	public String documentation;
	/** The parent block of this composite block. */
	public AdvanceCompositeBlock parent;
	/** The user-entered keywords for easier finding of this block. */
	public final List<String> keywords = Lists.newArrayList();
	/** The visual properties for the Flow Editor. */
	public final AdvanceBlockVisuals visuals = new AdvanceBlockVisuals();
	/** Contains the number of parameters for the varargs inputs. */
	public final Map<String, Integer> varargs = Maps.newHashMap();
	/**
	 * Load a block reference from an XML element which conforms the {@code flow-description.xsd}.
	 * @param source the element of a input or output
	 */
	@Override
	public void load(XElement source) {
		id = source.get("id");
		type = source.get("type");
		documentation = source.get("documentation");
		String kw = source.get("keywords");
		if (kw != null) {
			keywords.addAll(Strings.trim(Strings.split(kw, ',')));
		}
		visuals.load(source);
		for (XElement e : source.childrenWithName("vararg")) {
			varargs.put(e.get("name"), e.getInt("count"));
		}
	}
	@Override
	public void save(XElement destination) {
		destination.set("id", id);
		destination.set("type", type);
		destination.set("documentation", documentation);
		if (keywords.size() > 0) {
			destination.set("keywords", Strings.join(keywords, ","));
		} else {
			destination.set("keywords", null);
		}
		visuals.save(destination);
		for (Map.Entry<String, Integer> va : varargs.entrySet()) {
			XElement e = destination.add("vararg");
			e.set("name", va.getKey(), "count", va.getValue());
		}
	}
}

File: advance-api-cs/AdvanceAPIClient/Classes/Model/AdvanceBlockReference.cs
/*
 * Copyright 2010-2013 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml;

using AdvanceAPIClient.Core;
using AdvanceAPIClient.Interfaces;

namespace AdvanceAPIClient.Classes.Model
{
    /// <summary>
    /// The block reference for the {@code flow-description.xsd} used within composite blocks.
    /// </summary>
    public class AdvanceBlockReference : XmlReadWrite
    {
        /// <summary>
        /// Unique identifier of this block among the current level of blocks
        /// </summary>
        public string Id;
        /// <summary>
        /// block type identifier referencing a block description
        /// </summary>
        public string Type;
        /// <summary>
        /// user-entered documentation of this composite block.
        /// </summary>
        public string Documentation;
        /// <summary>
        /// parent block of this composite block
        /// </summary>
        public AdvanceCompositeBlock parent;
        /// <summary>
        /// user-entered keywords for easier finding of this block.
        /// </summary>
        public List<string> Keywords = new List<string>();
        /// <summary>
        /// visual properties for the Flow Editor
        /// </summary>
        public AdvanceBlockVisuals Visuals;
        /// <summary>
        /// Contains number of parameters for the varargs inputs
        /// </summary>
        public Dictionary<string, int> Varargs = new Dictionary<string, int>();


        /// <summary>
        /// Fill object data from xml node
        /// </summary>
        /// <param name="source"></param>
        protected override void LoadFromXmlNode(XmlNode source)
        {
 		    this.Id = GetAttribute(source, "id");
		    this.Type = GetAttribute(source, "type");
		    this.Documentation = GetAttribute(source, "documentation", null);
            this.Keywords = GetListAttribute(source, "keywords");
		    this.Visuals = CreateFromXml<AdvanceBlockVisuals>(source);
		    foreach (XmlNode node in GetChildren(source, "vararg"))
			    this.Varargs.Add(GetAttribute(node, "name"), GetIntAttribute(node, "count", 0));
		}


        /// <summary>
        /// Fills object's xml
        /// </summary>
        /// <param name="node">Empty node for the object with name "tagname"</param>
        protected override void FillXmlElement(XmlElement node)
        {
            AddAttribute(node, "id", this.Id);
            AddAttribute(node, "type", this.Type);
            AddAttribute(node, "documentation", this.Documentation);
            AddListAttribute(node, "keywords", this.Keywords);
            this.Visuals.AddToElement(node);
            foreach (KeyValuePair<string, int> va in this.Varargs)
            {
                XmlElement e = AddAttributeNode(node, "vararg", "name", va.Key);
                AddAttribute(e, "count", va.Value.ToString());
            }
        }
    }
}
